### PR TITLE
Refactor: Abstract message broker functionality

### PR DESCRIPTION
**Pull request by sphuber**
RabbitMQ is used as a message broker to allow processes to be
distributed across multiple daemon workers. Setting up RabbitMQ as a
service can be a hurdle for users and for simple use cases the
additional scaling it provides is not always necessary.

The idea is to provide a light version of AiiDA that doesn't require a
message broker, making it easy to setup and install to allow users to
quickly familiarize themselves with AiiDA before committing to a more
production ready setup.

Currently, the RabbitMQ functionality is a bit dispersed throughout the
codebase, making it difficult to make its use optional. Therefore, the
first step is to refactor the code and centralize all RabbitMQ specific
functionality.

The `aiida.brokers` module is introduced, which provides the abstract
class `Broker`. This is the interface that a message broker, such as
RabbitMQ should implement. The existing RabbitMQ implementation is added
to the `RabbitmqBroker` class. Note that this interface is extremely
preliminary and has not been fleshed out. The main purpose of
abstracting the broker at this point is to allow to run AiiDA without
broker, not necessarily to provide a complete interface ready to be
implemented by plugins for different broker services.

The `Manager` class now provides the `get_broker` method which will load
the broker instance based on the plugin that is defined by the profile.
This is an entry point that should be registered in the `aiida.brokers`
group. The broker is optional, in which case the profile can specify
`None` for the broker backend. In this case, `Manager.get_broker`
returns `None` so the caller should check explicitly for this case
before attempting to use the broker for a given profile.

These changes are just a first refactoring that don't change any
functionality and everything is backwards compatible. In the future,
functionality can be added to allow creating profiles that do not define
a broker. These can then be used to run processes locally in the same
Python interpreter in which they are launched.

### CODE REVIEW

**Comment by sphuber** [comment]
@edan-bainglass @unkcpz There is a problem with the tests you added for the `aiida-core-dev` Docker image recently. The problem is that it checks out the default branch of the main repository and so tests are run against that. However, they should really be run against the branch of the PR. I needed to change the tests since part of the functionality they test changes in this PR. How can we fix this? Can we make the remote and branch that is checked out for the tests configurable somehow?

**Comment by unkcpz** [comment]
Hi @sphuber, true, that is a problem. I think the proper way is instead of clone the repo, in the dockerfile we copy the aiida-core src to the container.

**Comment by unkcpz** [comment]
I open a PR to try to address the issue https://github.com/aiidateam/aiida-core/pull/6314

**Comment by sphuber** [comment]
> I open a PR to try to address the issue #6314, can you rebase this PR to test if it works?

Seems to do the trick, thanks! I will merge the other PR

**Comment by sphuber** [comment]
@unkcpz would you perhaps have the time to have a look at this PR? Don't worry about the separate commits. Just look at the total changes. I just added the second commit in case I need to revert, but I don't think we will. I think the current solution (without dummy `NoCommsBroker` plugin for profiles without a broker, and just have `Manager.get_broker()` return `None`) is better.

**Comment by giovannipizzi** [comment]
Really cool and useful, thanks @sphuber! Is there already documentation of the API of the Broker class, i.e. what functionality one needs to implement if I want to provide a different broker, what are the requirements of the Broker, etc?

**Comment by sphuber** [comment]
> Really cool and useful, thanks @sphuber! Is there already documentation of the API of the Broker class, i.e. what functionality one needs to implement if I want to provide a different broker, what are the requirements of the Broker, etc?

Not yet. This would require more work and I would prefer not to have that block this, because the most important reason for this work is to enable service-less profile. The fleshing out of the broker interface will require a lot more work. This is also the reason why I wouldn't officially document it _yet_, because likely we will have to improve/change this in the future once we figure out what the correct abstract interface should be.

**Comment by edan-bainglass** [comment]
> @edan-bainglass @unkcpz There is a problem with the tests you added for the `aiida-core-dev` Docker image recently. The problem is that it checks out the default branch of the main repository and so tests are run against that. However, they should really be run against the branch of the PR. I needed to change the tests since part of the functionality they test changes in this PR. How can we fix this? Can we make the remote and branch that is checked out for the tests configurable somehow?

I just tested on my end. I ran the docker tests on `main` (with @unkcpz's recent "fix" PR), and then with my `entity-fields` PR on top of it. In both cases, the tests failed with a `TCP connection closed`. However, simply rerunning the tests pushed through and finished successfully. Not great!

Note that for @sphuber, the tests appear to fail at the `aiida-core-dev` stage. However, in my case, they were already failing at the `aiida-core-base` stage, so this is perhaps a bigger issue. @unkcpz can we look into this? Also, @unkcpz, I'm unclear as to how your PR fixes the issue mentioned by @sphuber, i.e. tests running against the wrong branch. Is it not possible to fix this with the clone approach? Just checkout the PR branch pre-test stage?

**Comment by sphuber** [comment]
>  Is it not possible to fix this with the clone approach?

Then you would have to start figuring out in the script itself what fork and branch the PR is coming from. If that is even possible, it is unnecessarily complicated right, because you _have_ the correct code checked out locally by definition when you run the build.

> In both cases, the tests failed with a TCP connection closed. However, simply rerunning the tests pushed through and finished successfully. Not great!

Can you give more information on the exact error? Was there an exception? What is the stack trace? Where is the error coming from? And I see now on your PR that the build passed. Did you just restart the failed build, and the second time around it passed? Is that what you mean, that the problem seems transient?

**Comment by edan-bainglass** [comment]
> > Is it not possible to fix this with the clone approach?
> 
> Then you would have to start figuring out in the script itself what fork and branch the PR is coming from. If that is even possible, it is unnecessarily complicated right, because you _have_ the correct code checked out locally by definition when you run the build.

Thanks @sphuber for clarifying. I understand now how copying the local code (which is in the PR branch) fixes this issue.

> 
> > In both cases, the tests failed with a TCP connection closed. However, simply rerunning the tests pushed through and finished successfully. Not great!
> 
> Can you give more information on the exact error? Was there an exception? What is the stack trace? Where is the error coming from? 

I attempted to reproduce the issue a few times. Of course, it now passes every time. Nevertheless, here's an excerpt:

image

See the `role "root" does not exist` followed by a closing of the TCP connection? The test got stuck there. It just reattempted that part over and over until it timed out. At least that's what it seems to be doing.

> And I see now on your PR that the build passed. Did you just restart the failed build, and the second time around it passed? Is that what you mean, that the problem seems transient?

Yes, once I noticed that locally it failed, but then passed, I reran the test on my PR, which passed. No change introduced!

File: src/aiida/manage/manager.py [hunk 1]
```diff
@@ -272,6 +274,32 @@ def get_profile_storage(self) -> 'StorageBackend':
...
+            raise ConfigurationError(
+                'Could not determine the current profile. Consider loading a profile using `aiida.load_profile()`.'
+            )
+
+        if self._broker is None and self._profile.process_control_backend is not None:
+            from aiida.plugins import BrokerFactory
+
+            entry_point = self._profile.process_control_backend
+
+            # Backwards compatibility. Before adding broker entry points, profiles used to define ``rabbitmq``.
```
Comment (unkcpz) [inline src/aiida/manage/manager.py:294]: I didn't get this, where was the ``rabbitmq`` originally defined?

**Review by unkcpz — CHANGES_REQUESTED** [review]
Very cool, thanks @sphuber! Apparently you find another layer of abstraction. 
I only have two small questions, if you can further explain those.

File: src/aiida/brokers/broker.py [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+    def __init__(self, profile: 'Profile') -> None:
+        """Construct a new instance.
+
+        :param profile: The profile.
+        """
+        self._profile = profile
+
+    @abc.abstractmethod
+    def get_communicator(self):
+        """Return an instance of :class:`kiwipy.Communicator`."""
```
[the remark this answers is not in the document]
Comment (sphuber) [reply src/aiida/brokers/broker.py:23]: For now, a new broker implementation would indeed have to implement the `kiwipy.Communicator` interface. For now, I just did the minimal changes necessary to abstract the broker. This is not so much to make it possible to provide other broker implementations, but really to make it easy to run AiiDA with _no_ broker at all. I thought that abstracting the functionality in this way made it as easy as possible to accomplish that.

As I mentioned [in this comment](https://github.com/aiidateam/aiida-core/pull/6303#issuecomment-1997839409) actually fully abstracting the broker/communicator interface would take more work, but is not the main goal here.

**Comment by sphuber** [comment]
> Very cool, thanks @sphuber! Apparently you find another layer of abstraction.

Hehe yeah 😅 I was really hoping not to have to do that, but it seems the best option to allow running AiiDA in minimal mode that doesn't require RabbitMQ.

**Review by unkcpz — APPROVED** [review]
Looks all good, I'll give it a more close test on PR `verdi blitz`.

**Pull request merged**